public/app/features/scenes/core/types.ts

```diff
@@ -0,0 +1,122 @@
+import React from 'react';
+import { Subscribable } from 'rxjs';
+
+import { EventBus, PanelData, TimeRange, UrlQueryMap } from '@grafana/data';
+
+import { SceneVariableSet } from '../variables/types';
+
+export interface SceneObjectState {
+  key?: string;
+  size?: SceneObjectSize;
+  $timeRange?: SceneTimeRange;
+  $data?: SceneObject<SceneDataState>;
+  $editor?: SceneEditor;
+  $variables?: SceneVariableSet;
+}
+
+export interface SceneObjectSize {
+  width?: number | string;
+  height?: number | string;
+  xSizing?: 'fill' | 'content';
+  ySizing?: 'fill' | 'content';
+  x?: number;
+  y?: number;
+  minWidth?: number | string;
+  minHeight?: number | string;
+}
+
+export interface SceneComponentProps<T> {
+  model: T;
+  isEditing?: boolean;
+}
+
+export type SceneComponent<TModel> = React.FunctionComponent<SceneComponentProps<TModel>>;
+
+export interface SceneDataState extends SceneObjectState {
+  data?: PanelData;
+}
+
+export interface SceneObject<TState extends SceneObjectState = SceneObjectState> extends Subscribable<TState> {
+  /** The current state */
+  state: TState;
+
+  /** True when there is a React component mounted for this Object */
+  isMounted?: boolean;
+
+  /** SceneObject parent */
+  parent?: SceneObject;
+
+  /** For only only used by from root to broadcast events */
+  events: EventBus;
+
+  /** Utility hook that wraps useObservable. Used by React components to subscribes to state changes */
+  useState(): TState;
+
+  /** How to modify state */
+  setState(state: Partial<TState>): void;
+
+  /** Utility hook for main component so that object knows when it's mounted */
+  useMount(): this;
+
+  /** Called when component mounts. A place to register event listeners add subscribe to state changes */
+  onMount(): void;
+
+  /** Called when component unmounts. Unsubscribe to events */
+  onUnmount(): void;
```

Comment:
Not super happy with these names, thinking of changing them to either 
* onInit and onCleanUp and isInitialized 

But the problem is that an Object be mounted & unmounted and re-mounted again (for example if you come back to the same dashboard and we have cached it, or if we have a tabbed component or rows component and some objects where temporary collapsed from view, then re-expanded) 

Maybe
* onSetup and onTeardown and isActive
or
* activate and deactivate and isActive